File: r/tests/testthat/test-render-svg-highlight.R
.toy_3set_for_highlight <- function() {
    ds <- methods::new("VennDataset",
        set_names = c("A", "B", "C"),
        items = list(A = c("x", "y"), B = c("y", "z"), C = c("z", "w")),
        item_order = c("x", "y", "z", "w"),
        universe_size = 10L, source_path = NULL, format = "csv")
    analyze(ds)
}

test_that("render_venn_svg with highlight=NULL is byte-identical to no-highlight default", {
    res <- .toy_3set_for_highlight()
    svg_default <- render_venn_svg(res)
    svg_null    <- render_venn_svg(res, highlight = NULL)
    expect_equal(svg_default, svg_null)
})

test_that("render_venn_svg highlight=AB desaturates shape C (set NOT in any highlighted region)", {
    res <- .toy_3set_for_highlight()
    svg <- render_venn_svg(res, highlight = "AB")
    # ShapeC fill should be rewritten to the highlight grey.
    expect_match(svg, 'id="ShapeC"[^>]*style="[^"]*fill:#cccccc')
    # ShapeA / ShapeB fills should NOT be rewritten (they are in the
    # highlighted region AB).
    expect_no_match(svg, 'id="ShapeA"[^>]*style="[^"]*fill:#cccccc')
    expect_no_match(svg, 'id="ShapeB"[^>]*style="[^"]*fill:#cccccc')
})

test_that("render_venn_svg highlight accepts an integer bitmask vector", {
    res <- .toy_3set_for_highlight()
    # Bitmask 3 = AB, bitmask 5 = AC. So A is in both, B in mask 3, C in mask 5.
    # Highlighted union mask = 7 (A|B|C). No set should be desaturated.
    svg <- render_venn_svg(res, highlight = c(3L, 5L))
    expect_no_match(svg, 'id="Shape[ABC]"[^>]*style="[^"]*fill:#cccccc')
})

test_that("render_venn_svg highlight composes with parse_region_expression", {
    res <- .toy_3set_for_highlight()
    masks <- parse_region_expression("A & B", n_sets = 3L)   # c(3L, 7L)
    svg <- render_venn_svg(res, highlight = masks)
    # Union of regions 3 and 7 covers A, B, C -> no set desaturated.
    expect_no_match(svg, 'id="Shape[ABC]"[^>]*style="[^"]*fill:#cccccc')

    # But A & B with C excluded:
    masks2 <- parse_region_expression("A & B & ~C", n_sets = 3L)  # 3L only
    svg2 <- render_venn_svg(res, highlight = masks2)
    # Only A and B contribute -> C is desaturated.
    expect_match(svg2, 'id="ShapeC"[^>]*style="[^"]*fill:#cccccc')
})

test_that("render_venn_svg highlight rejects unknown region labels", {
    res <- .toy_3set_for_highlight()
    expect_error(render_venn_svg(res, highlight = "XYZ"),
                 "unknown region label")
})

test_that("render_venn_svg highlight rejects unknown bitmasks", {
    res <- .toy_3set_for_highlight()
    expect_error(render_venn_svg(res, highlight = 99L),
                 "no region with mask")
})

test_that("render_venn_svg highlight rejects type-mismatched inputs", {
    res <- .toy_3set_for_highlight()
    expect_error(render_venn_svg(res, highlight = TRUE),
                 "must be NULL")
    expect_error(render_venn_svg(res, highlight = list(1L, 2L)),
                 "must be NULL")
})

File: python/tests/test_render_svg_highlight.py
"""Tests for the highlight spotlight mode in render_venn_svg."""

from __future__ import annotations

import re

import pytest

from venn_diagram_lab import (
    Dataset,
    analyze,
    parse_region_expression,
    render_venn_svg,
)


def _toy_3set():
    ds = Dataset(
        set_names=["A", "B", "C"],
        items={"A": {"x", "y"}, "B": {"y", "z"}, "C": {"z", "w"}},
        source_path=None,
        format="csv",
        item_order=("x", "y", "z", "w"),
        universe_size=10,
    )
    return analyze(ds)


def test_highlight_none_is_identical_to_default():
    res = _toy_3set()
    svg_default = render_venn_svg(res).svg
    svg_null = render_venn_svg(res, highlight=None).svg
    assert svg_default == svg_null


def test_highlight_AB_desaturates_set_C():  # noqa: N802 -- AB/C refer to set letters
    res = _toy_3set()
    svg = render_venn_svg(res, highlight=["AB"]).svg
    assert re.search(r'id="ShapeC"[^>]*style="[^"]*fill:#cccccc', svg)
    assert not re.search(r'id="ShapeA"[^>]*style="[^"]*fill:#cccccc', svg)
    assert not re.search(r'id="ShapeB"[^>]*style="[^"]*fill:#cccccc', svg)


def test_highlight_accepts_bitmask_list():
    res = _toy_3set()
    # Mask 3 = AB, mask 5 = AC. Union covers A, B, C => no desaturation.
    svg = render_venn_svg(res, highlight=[3, 5]).svg
    assert not re.search(r'id="Shape[ABC]"[^>]*style="[^"]*fill:#cccccc', svg)


def test_highlight_composes_with_parse_region_expression():
    res = _toy_3set()
    # "A & B" -> [3, 7]. Union covers A, B, C => no desaturation.
    masks = parse_region_expression("A & B", n_sets=3)
    svg = render_venn_svg(res, highlight=masks).svg
    assert not re.search(r'id="Shape[ABC]"[^>]*style="[^"]*fill:#cccccc', svg)

    # "A & B & ~C" -> [3]. Only A, B contribute => C desaturated.
    masks2 = parse_region_expression("A & B & ~C", n_sets=3)
    svg2 = render_venn_svg(res, highlight=masks2).svg
    assert re.search(r'id="ShapeC"[^>]*style="[^"]*fill:#cccccc', svg2)


def test_highlight_rejects_unknown_label():
    res = _toy_3set()
    with pytest.raises(ValueError, match="unknown region label"):
        render_venn_svg(res, highlight=["XYZ"])


def test_highlight_rejects_unknown_bitmask():
    res = _toy_3set()
    with pytest.raises(ValueError, match="no region with mask"):
        render_venn_svg(res, highlight=[99])


def test_highlight_rejects_type_mismatched_inputs():
    res = _toy_3set()
    with pytest.raises(ValueError, match="must be None"):
        render_venn_svg(res, highlight=[True, False])
